blog: Running Wasm Components From the Command Line (approved) (#126)
bytecodealliance/bytecodealliance.org: _posts/2025-04-10-invoking-component-functions-in-wasmtime-cli.md

Author: Thanks for the review Victor (@vados-cosmonic).
I have pushed a new commit that covers about half of the feedback. I will carve out some time tonight or tomorrow to get to the rest.

Author: Great suggestions, BTW. Thanks a bunch!

Reviewer: No worries! Glad I could help :)

```bash
wasmtime run --invoke "get-answer()" target/wasm32-wasip1/debug/wasm_answer.wasm
```

## Wasm Value Encoding (WAVE)

Invoke leverages [WAVE](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasm-wave/README.md) as a standard way to encode function calls and/or results. WAVE is a human-oriented text encoding of Wasm Component Model values and is designed to be consistent with the [WIT IDL format](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). Here are a few pointers for constructing your `wasmtime run --invoke` commands using WAVE.

## Parentheses

Parentheses after the exported function's name are mandatory. The presence of the parenthesis `()` signifies function invocation, as opposed to the function name just being referenced.

## Double Quotes

The exported function's name and mandatory exported function's parentheses must all be enclosed in one set of double quotes, i.e. `"get-answer()"`.

The result from our correctly typed command above is as follows:

```bash
42
```

Note: If your function takes a string argument, ensure that you use escaped double quotes inside the parentheses. For example:

```bash

> wasmtime run - invoke "initialize(\"hello\")" foo.wasm

Reviewer 2: ```suggestion
wasmtime run - invoke 'initialize("hello")' foo.wasm
```

Using single quotes on the outside avoids the need to escape inner double quotes.

> wasmtime run - invoke "initialize(\"hello\")" foo.wasm

Reviewer 2: Here and below `s/- invoke/--invoke/`

Author: Ah, thank you. I will fix and commit now.

---
title: "Running WebAssembly (Wasm) Components From the Command Line"
author: "Tim McCallum"
date: "2025-04-10"
github_name: "tpmccallum"
excerpt_separator: <!--end_excerpt-->
---

> Wasmtime now supports invoking Wasm component exports directly from the command line with the new `--invoke` flag. This article walks through building a Wasm component in Rust, writing a WIT interface, and using `wasmtime run --invoke` to execute specific functions (enabling powerful workflows for scripting, testing, and integrating Wasm into modern development pipelines).

Reviewer: ```suggestion
Wasmtime now supports invoking Wasm component exports directly from the command line with the new `--invoke` flag. 
This article walks through building a Wasm component in Rust, writing a WIT interface, and using `wasmtime run --invoke` to execute specific functions (enabling powerful workflows for scripting, testing, and integrating Wasm into modern development pipelines).
```

What do you think about some separation here? Also I wonder if Wasmtime or "supports invoking wasm component exports" deserves a link to somewhere (a PR might be too low level... the issue?)

Also, what do you think about making this a bulleted list? Might be a little easier to parse than the paragraph form

<!--end_excerpt-->

## The Evolution of Wasmtime's CLI

Wasmtime's `run` subcommand has traditionally excelled at running Wasm modules, whether in binary (`.wasm`) or text (`.wat`) format. In this article, we will create a Wasm component that exports a function and then demonstrate how to invoke the function using `wasmtime run --invoke`.

> ## Housekeeping

Reviewer: What do you think about naming this something more related to "dependencies" or "required tooling" or "setup"?

Author: Thanks, Victor. I will update locally and push these changes. 
Great suggestions, thank you.

Author: Updated to "Tooling & Dependencies"

If you want to follow along, please install:

* [Rust](https://www.rust-lang.org/tools/install) (if you already have Rust installed, make sure you are on [the latest version](https://github.com/rust-lang/rust/releases) using `rustup update`),

> * Cargo Component via the `cargo install cargo-component` command (if already installed, please make sure you are on [the latest version](https://github.com/bytecodealliance/cargo-component/releases)), and

Reviewer: ```suggestion
* [`cargo component`](https://crates.io/crates/cargo-component) via the `cargo install cargo-component` command (if already installed, please make sure you are on [the latest version](https://github.com/bytecodealliance/cargo-component/releases)), and
```

* [Wasmtime](https://docs.wasmtime.dev/cli-install.html) or download a [Wasmtime precompiled binary](https://docs.wasmtime.dev/cli-install.html#download-precompiled-binaries). (If you already have wasmtime installed, please make sure you are using [the latest version](https://github.com/bytecodealliance/wasmtime/releases).)

You can check versions using the following commands:

> ```bash

Reviewer: personal preference, but I like to use `bash` for bash scripts and `console` for things that should be entered at a console.

Author: Sure thing, I am happy with that preference.
Updating in the next commit.

wasmtime --version
cargo component --version
rustc --version
```

## New Library

Let's start by creating a new Rust library that we will later convert to a Wasm component using `cargo component`:

```bash

> cargo new --lib wasm-answer
> cd wasm-answer

Reviewer: ```suggestion
cargo component new --lib wasm-answer
cd wasm-answer
```

Author: Thank you, updated in next commit.

```

Install `wit-bindgen`:

```bash
cargo add wit-bindgen

> ```

Reviewer: If people use cargo component this isn't necessary -- `cargo component` now has a `cargo component bindings` command that generates bindings (and uses the same libs under the covers of course)

Author: Thanks, I have updated this in the next commit.

If you open the `Cargo.toml` file, you will notice that the `wit-bindgen` dependency has been automatically added for us (by the above `cargo add command`). While you have the `Cargo.toml` open, please go ahead and add the following:

```toml
[lib]
crate-type = ["cdylib"]
```

The `Cargo.toml` file will look something like the following:

```toml
[package]
name = "wasm-answer"
version = "0.1.0"
edition = "2024"

[lib]
crate-type = ["cdylib"]

[dependencies]
wit-bindgen = "0.41.0"
```

We need to set a special target that allows us to leverage the Wasm component model via `cargo component`. So please go ahead and add the Wasm target shown below and also set your default Rust toolchain to `nightly`:

```bash
rustup toolchain install nightly
rustup default nightly

> rustup target add wasm32-wasip1

Reviewer: Is this true? You shouldn't *need* nightly to use `cargo component`, and to produce `wasm32-wasip1` binaries either

Author: Good spotting. I was able to get it to work without nighly if I used all of the latest versions mentioned in the tooling and dependencies section. Updating this in the next commit.

```

## WIT

Now, let's create a hand-written `.wit` file. Open a new file `answer.wit` in and new directory `wit`:

```bash
mkdir wit
vi wit/answer.wit

> ```

Reviewer: If we're going to build WIT from scratch, what do you think about including this *first* along with a slight description of the syntax/what's happening?

Creating a component and then creating WIT works, but a WIT-first approach I think fits the ideal usage -- people thinking in terms of intended functionality/structure first then creating bindings that satisfy.

Author: I am taking your advice from above and creating the project using cargo component (and so the WIT will already be created for us). I am just adding a section that shows how to update the WIT in the next commit.

Add the following content to the `answer.wit` file:

```wit
package example:answer;
world answer-world {
    export get-answer: func() -> u32;
}
```

Your directory structure should look like the following:

```bash

> tree .

Reviewer: ```suggestion
tree wasm-answer
```

Doing the folder might be a bit more descriptive here (and maybe you could leave off the actual command and just have the tree?

Author: Good point.

.
├── Cargo.lock
├── Cargo.toml
├── src
│   └── lib.rs
└── wit
    └── answer.wit
```

Next, we write the logic that supports our WIT. Open the `src/lib.rs` file and add the following content:

```rust
wit_bindgen::generate!({
    world: "answer-world",  
    path: "wit/answer.wit",
});

> 

Reviewer: If we're using `cargo component` this looks a bit different -- `cargo component` manages the bindings and writes them out to disk into a `bindings.rs` (which is a bit easier to inspect for users, IMO it's pretty useful, though power users will prefer `wit_bindgen::generate` & debugging w/ `cargo expand`).

If we'd like to use `wit_bindgen::generate` then IMO we can ignore `cargo component`, but I think we should push people to use `cargo component`.

Also, as a side note, I think the pattern of encapsulating the generated output of `wit_bindgen` is really helpful for readability. Something like the following for the code in here:

```rust
mod bindings {
  wit_bindgen::generate(...);
  
  export
}

// ...
use bindings::example::answer::answer_world;

/// Target for WIT world implementation
///
/// Generated rust bindings for the `answer` WIT world used in this project
/// will have traits that are implemented by this struct.
struct Component;

impl answer_world::Guest for Component {
    // ...
}
```

Author: Thanks, I agree that sticking with `cargo component` will help people quickly try out this introductory example. In relation to code snippets, I have stuck to pasting (into the article) what is generated using the `cargo component` approach (`.rs`, `.wit`). Also, did not include any of the generated Rust bindings in the snippets because I think what we have tells a good enough story for this article's purpose.

struct Answer;
impl Guest for Answer {
    fn get_answer() -> u32 {
        42
    }
}
export!(Answer);
```

Now, let's create the Wasm component with the exported function:

```bash

> cargo component build --target wasm32-wasip1

Reviewer: ```suggestion
cargo component build
```

IIRC target explicitly isn't necessary here, also the output of cargo component *does* make it clear where the file is saved to, so it might be worth including that output here, but meh maybe not

Author: I put a bit of thought into this, and I think that explicitly stating the target is a good idea. I might keep this in if that's ok with you.

Reviewer: That sounds good to me -- thanks for considering it!

```

If we take another look at our directory structure, we will see that cargo component has automatically generated `bindings.rs` and that we now have a target directory that contains `wasm32-wasip1` path. 

```bash

> tree . -L 2

Reviewer: ```suggestion
tree . -L 2
```
IMO you can leave the tree command out

Author: Thanks, removed this one.

.
├── Cargo.lock
├── Cargo.toml
├── src
│   ├── bindings.rs
│   └── lib.rs
├── target
│   └── wasm32-wasip1
└── wit
    └── answer.wit
```

Our newly generated `.wasm` file now lives at the following location:

```bash
ls target/wasm32-wasip1/debug/wasm_answer.wasm 

> ```

Reviewer: Another nice thing to show here might be the output of `file`:

```
file target/wasm32-wasip1/debug/wasm_answer.wasm
target/wasm32-wasip1/debug/wasm_answer.wasm: WebAssembly (wasm) binary module version 0x1000d
```

Also, we might want to show that we can build release versions with `--release`, because debug is 2MB and release is 64K.

Author: Great idea, I will add that in the next commit.

## Default Entry Point vs. Exported Function

> If we were not using the component model and just creating binary executable, we could use a simple command like `wasmtime run foo.wasm` to execute our code.

Reviewer: I'm not sure this is enough of an explanation of what's happening. Wasm has the concept of reactor modules and command modules, which is actually *also* subtly distinct from whether or not a component exports the `wasi:cli/run` interface (roughly the distinction is that `command` modules have a `_start` export IIRC, but you can technically export `wasi:cli/run` from *either* a reactor or command module!).

I don't know how much could reasonably be included here without blowing up the section (and making this post too technical) but IMO mentioning `wasi:cli/run` and the standardization there might be worth it.

Author: Thanks @vados-cosmonic 
I have updated this in the next commit/push.

> But in this case, we are making the point that Wasm has evolved beyond simple modules to embrace the Wasm Component Model. Developers need more granular control over the execution and composition of components. The benefits are code reuse and cross-language sharing. For example, Rust, Javascript, and Python logic can all interoperate when the explicit interfaces are defined using WIT. For example, [seamlessly interoperable compression tasks across native C++, Rust, and also in a Wasm runtime](https://medium.com/wasm/wasm-component-model-seamless-compression-c-rust-and-wasm-3b8d52ed8b31).

Reviewer: This is a great point to include, but IMO it's a bit different/unrelated to entrypoints and the idea of a binary executable -- here we're discussing the ability to share/reuse functionality which is firmly in the as-a-lib camp.

IMO this is inline with the example so far, because we *do* create a `--lib` component (i.e. a "reactor") which is sharable/reusable -- so maybe this is something that belongs way up top (maybe where we're describing the WIT? to emphasize WIT's capabilities as a network-independent abstraction/IDL for computation)

Author: Thanks, I have fleshed this out to be clearer, using headings and separating the concept of a module vs a component.
Will push changes soon.

## Invoking Exported Functions

> The addition of the `--invoke` feature (within the `wasmtime run` subcommand) allows users to specify an exported function to execute, complete with arguments, directly from the command line interface (CLI) terminal. Being able to invoke functions in the CLI is a huge step forward for not only executing Wasm via consoles or Bash scripting but will also play a big role in testing, debugging, and integrating Wasm components into workflows without having to embed them into a host application during development.

Reviewer: This is a good description but could maybe use a caveat noting that up until now `wasmtime` could only reasonably invoke things that exported `wasi:cli/run`, *because* interfaces must be known ahead of time (and `wasi:cli/run` is standardized and used exactly for that purpose). `--invoke` makes it possible to execute *arbitrary* exports.

This is almost like being able to invoke any OpenAPI endpoint or any gRPC endpoint from the command line rather than some pre-established entrypoint.

Author: Thanks again, changes in next push.

> ## How Invoke Works: A Practical Example

Reviewer: Maybe this section should go right after the end the "building" part of the instructions? How to build what we just created *then* going through how things work under the covers IMO flows a bit smoother

Author: Thanks, have moved this around and edited accordingly.

Originally, the `wasmtime run` command would take one positional argument (the name of the module) and just run that `.wasm` file:

```bash
wasmtime run foo.wasm
```

The `wasmtime run` command now accepts an optional `--invoke` argument, which can execute the name of an exported function that resides in the (`.wasm`) module:

```bash
wasmtime run --invoke 'get-answer()' target/wasm32-wasip1/debug/wasm_answer.wasm
```

## Wasm Value Encoding (WAVE)

> Invoke leverages [WAVE](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasm-wave/README.md) as a standard way to encode function calls and/or results. WAVE is a human-oriented text encoding of Wasm Component Model values and is designed to be consistent with the [WIT IDL format](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). Here are a few pointers for constructing your `wasmtime run --invoke` commands using WAVE.

Reviewer: It might be worth motivating this a bit more *before* noting that we use WAVE -- laying out that the problem is that we have to be able to transfer *arbitrary* data (i.e. almost like representing JSON on the command line, without typing verbose JSON objects) to be able to pass it in to a CLI.

Author: Great suggestion, I have added a pre-paragraph that introduces the problem that wave solves, then on to the expxlaination of wave.

## Parentheses

Parentheses after the exported function's name are mandatory. The presence of the parenthesis `()` signifies function invocation, as opposed to the function name just being referenced.

## Quotes

The exported function's name and mandatory exported function's parentheses must all be enclosed in one set of single quotes, i.e. `'get-answer()'`.

> The result from our correctly typed command above is as follows:

Reviewer: IMO these could both be notes/info boxes closer to the actual usage example?

Author: Thanks, have moved content around and provided segways into examples.

```bash
42
```

If your function takes a string argument, ensure that you envelop your string in double quotes (inside the parentheses). For example:

```bash
wasmtime run - invoke 'initialize("hello")' foo.wasm
```

**Please note:** If you enclose your function call using double quotes, your string argument will require its double quotes to be escaped (escaping quotes is more complicated and harder to read and therefore not ideal). For example:

```bash
wasmtime run - invoke "initialize(\"hello\")" foo.wasm
```

Lastly, if your exported function takes more than one argument, you will need to separate each argument with a single comma `,` as shown below:

```bash
wasmtime run -- invoke 'initialize("Pi", 3.14)' foo.wasm
wasmtime run -- invoke 'add(1, 2)' foo.wasm
```

With the ability to invoke Wasm component exports directly from the command line, developers unlock powerful workflows:

> 

Reviewer: IMO this *could* go into new section! Exploring the possible use cases is super fun -- and maybe that could be worth even moving to the top. 

So the overall flow could be something like:

- Introduce new feature (identifying that now `wasmtime run` can run *arbitrary* component exports)
- Describe the possible applications
- Jump into an example
- Explain how the example works and take a peek behind the curtain

Author: Thanks, have rewritten this section.

Author: Didn't ellaborate on an additional example. Thinking that could be another future article. Keep this one succinct about invoke for components.

## Tooling & Dependencies

If you want to follow along, please install:

* [Rust](https://www.rust-lang.org/tools/install) (if you already have Rust installed, make sure you are on [the latest version](https://github.com/rust-lang/rust/releases)),
* [`cargo`](https://crates.io/crates/cargo) (if already installed, please make sure you are on [the latest version](https://crates.io/crates/cargo)),
* [`cargo component`](https://crates.io/crates/cargo-component) (if already installed, please make sure you are on [the latest version](https://crates.io/crates/cargo-component)), and
* [`wasmtime` CLI](https://docs.wasmtime.dev/cli-install.html) (or use a [precompiled binary](https://docs.wasmtime.dev/cli-install.html#download-precompiled-binaries)). If already installed, ensure you are using [the latest version](https://github.com/bytecodealliance/wasmtime/releases).

You can check versions using the following commands:

```console
$ rustc --version
$ cargo --version
$ cargo component --version

> $ wasmtime --version

Reviewer: If we're showing only commands, we should remove the `$` prefix, because it makes the commands hard to copy and paste into a terminal (ditto for all the other uses of `$`).

IMO it does make sense to include `$` in code blocks where you show a command and it's output (so users can differentiate between the two)

Author: Yeah, that's a really great point. I am keeping consistent with the `cli-options` page on [the wasmtime documentation](https://github.com/bytecodealliance/wasmtime/blob/main/docs/cli-options.md), so I guess we should leave them in for consistency regarding this `run` feature.

Reviewer: yeah I agree  we should be consistent here. It's a shame the wasmtime docs do that -- will try to find some time to send a PR and get to the bottom of why/if there's a good reason!

```

> For `cargo component` to generate a Wasm binary (`.wasm` file) that is compatible with the WASI Preview 1 standard, we must explicitly `add` the `wasm32-wasip1` target. This ensures that our component adheres to WASI’s system interface for non-browser environments (e.g., file system access, networking):

Reviewer: ```suggestion
For `cargo component` to generate a Wasm binary (`.wasm` file, in this case a Wasm module) that is compatible with the [WASI Preview 1 standard](https://github.com/WebAssembly/WASI/tree/main/legacy/preview1), we must explicitly `add` the `wasm32-wasip1` target. This ensures that our component adheres to WASI’s system interface for non-browser environments (e.g., file system access, networking):
```

Reviewer: Unfortunately, this change kind of exposes an issue -- technically we're not using a WebAssembly "component" here -- we're actually dealing with a Wasm module which adheres to WASI preview1.

It's a fine technical point, but the WebAssembly "binary" produced when targeting `wasm32-wasip1` is actually a WebAssembly *module* (i.e. the same stuff that you produce with `wasm32-unknown-unknown`), but adhering to/including a set of imports that represents WASI functionality. 

To convert a P1 module into a "component", we can do one of at least two things:

- Adapt (automatic, always works) with an adapter module (see wasmtime releases, `wasm-tools component embed`/`wasm-tools component new`)
- Target [`wasm32-wasip2` from Rust](https://blog.rust-lang.org/2024/11/26/wasip2-tier-2.html) (which does the adaptation for you, essentially)

I think there are a lot of ways you could update this post to reflect this; changing the focus on components to be focus on "WASI functionality", or updating the concept/description to be more technically accurate (and targeting `wasm32-wasip2`), etc -- and I'm happy with any way you'd like to change that if you would!

Author: Great point. I will update so that we use `wasm32-wasip2` throughout the article.

Author: The latest push uses `wasm32-wasip2` exclusively. 
The wording has been updated to mention **module** when referring to no component work and **component** when referring to any new builds that target `wasm32-wasip2`.

```console
$ rustup target add wasm32-wasip1
```

## New Library

> Let's start by creating a new Rust library that we will later convert to a Wasm component using `cargo component`:

Reviewer: Not a super strongly held opinion but I think this section might be better named something like "Creating a new WebAssembly component with Rust" or something like that, and then clarifying under the `cargo component new --lib` that we're creating a "library" project in Rust parlance which turns into a "reactor" or a "library" component on the WebAssembly side.

It's not like we need to get into what reactor modules versus command modules are here, but I think "Library" doesn't mean much to newcomers to start with, and it *does* mean something to the Rust-attuned newcomer, and then it means a third thing to the WebAssembly-curious.

Author: Thanks, updated in next push.

The `wit-bindgen-rt` dependency (with the `["bitflags"]` feature) under `[dependencies]`, and the `crate-type = ["cdylib"]` setting under the `[lib]` section.

Your `Cargo.toml` should now include these entries (as shown in the example below):

```toml
[package]
name = "wasm_answer"
version = "0.1.0"
edition = "2024"

[dependencies]
wit-bindgen-rt = { version = "0.41.0", features = ["bitflags"] }

[lib]
crate-type = ["cdylib"]

[package.metadata.component]
package = "component:wasm-answer"

[package.metadata.component.dependencies]
```

The directory structure of the `wasm_answer` example is automatically scaffolded out for us by `cargo component`:

```console
$ tree wasm_answer

wasm_answer
├── Cargo.lock
├── Cargo.toml
├── src
│   ├── bindings.rs
│   └── lib.rs
└── wit
    └── world.wit
```

Next, we add a `get_answer` function in the `src/lib.rs` file:

```rust
#[allow(warnings)]
mod bindings;

use bindings::Guest;

struct Component;

impl Guest for Component {
    fn get_answer() -> u32 {
        42
    }
}

bindings::export!(Component with_types_in bindings);
```

## WIT

> Now, we need to open the `.wit` file and slightly modify it for our use case:

Reviewer: It might be a good idea to put in a note about how `cargo component` generated a WIT file for us, but WIT files are actually core/central to component development -- they dictate the interface of the binary that is produced. 

Most design/creation of components should actually be WIT-first (i.e. thinking of the interface before building the actual component).

Author: Updates in next push.

```console
$ vi wit/world.wit

> ```

Reviewer: I'm not sure we should expect the average user to be familiar with `vi` or have it installed/accessible. I think letting people open the file however they want (without offering a suggestion) is a reasonable thing to expect a reader to be able to do without an explicit copy-pastable snippet (though it is helpful)

Author: Sure, sounds fair :)

Add the following content to the `answer.wit` file:

```wit
package component:wasm-answer;

world example {

>     export get-answer: func() -> u32;

Reviewer: Let's use an interface here -- this is a change we made in the component docs that IMO we should adopt here. Directly exporting functions *is* supported, but almost all of the time, people should be wrapping related functions and types in interfaces.

As an example see the [adder world in component-docs](https://github.com/bytecodealliance/component-docs/blob/main/component-model/examples/tutorial/wit/adder/world.wit) (it previously used a plain function export in a different files, but all guides there were updated to the interface style)

Author: 100%
Great idea, Victor ( @vados-cosmonic )
The latest push has the updated code that uses the interface style you recommended.
Good call.

```console
$ file target/wasm32-wasip1/debug/wasm_answer.wasm
target/wasm32-wasip1/debug/wasm_answer.wasm: WebAssembly (wasm) binary module version 0x1000d
```

We can also use the `--release` option which optimises builds for production:

```console
$ cargo component build --target wasm32-wasip1 --release
```

If we check the sizes of the `debug` and `release`, we see a difference of `1.9M` and `16K`, respectively.

Debug:

```console
$ du -mh target/wasm32-wasip1/debug/wasm_answer.wasm
1.9M	target/wasm32-wasip1/debug/wasm_answer.wasm
```

Release:

```console
$ du -mh target/wasm32-wasip1/release/wasm_answer.wasm
16K	target/wasm32-wasip1/release/wasm_answer.wasm
```

## How Invoke Works: A Practical Example

The `wasmtime run` command can take one positional argument and just run a `.wasm` or `.wat` file:

```console
$ wasmtime run foo.wasm
$ wasmtime run foo.wat
```

### Invoke: Wasm Modules

In the case of a Wasm **module** that exports a raw function directly, the `run` command accepts an optional `--invoke` argument, which is the name of an exported raw function (of the **module**) to run:

```sh
$ wasmtime run --invoke initialize foo.wasm
```

### Invoke: Wasm Components

> In the case of a Wasm **component** that uses typed interfaces (defined by [the component model](https://component-model.bytecodealliance.org/design/components.html)), the `run` command now also accepts the optional `--invoke` argument for calling an exported function of a **component**. 

Reviewer: ```suggestion
In the case of a Wasm **component** that uses typed interfaces (defined [in WIT](https://component-model.bytecodealliance.org/design/wit.html), in concert with [the component model](https://component-model.bytecodealliance.org/design/components.html)), the `run` command now also accepts the optional `--invoke` argument for calling an exported function of a **component**. 
```

Author: I have pasted your version in my local copy and it will appear in the next push.
Thanks - this is a great addition.

> However, the calling of an exported function of a **component** uses [WAVE](https://github.com/bytecodealliance/wasm-tools/tree/a56e8d3d2a0b754e0465c668f8e4b68bad97590f/crates/wasm-wave#readme)(a human-oriented text encoding of Wasm Component Model values). For example:

Reviewer: Not very opinionated here but what do you think about noting that WAVE is necessary since exported WIT functions can take a large (effectively arbitrary) amount of inputs? Maybe just a link down to the bottom section since this is covered later?

Author: I might keep the explanation of WAVE to a minimum because I have not had a great deal of time to test it out as much as I would like.

Reviewer: Yeah I can totally understand that!

> ```sh

Reviewer: NIT: inconsistent use of `sh` here versus `console`

Author: Thank you!

$ wasmtime run --invoke 'initialize()' foo.wasm
```

> You will notice the different syntax of `initialize` versus `initialize()` when referring to a **module** versus a **component**, respectively.

Back to our `get-answer()` example:

```console
$ wasmtime run --invoke 'get-answer()' target/wasm32-wasip1/debug/wasm_answer.wasm
```

> 

Reviewer: It might be a good idea to include the expected output here, like you did below

Author: Done.

You will notice that the above `get-answer()` function call does not pass in any arguments. Let's discuss how to represent the arguments passed into function calls in a structured way (using WAVE).

#### Wasm Value Encoding (WAVE)

> Transferring and invoking complex argument data via the command line is challenging, especially with Wasm components that use diverse value types. To simplify this, Wasm Value Encoding ([WAVE](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasm-wave/README.md)) was introduced; offering a concise way to represent structured values directly in the CLI. WAVE provides a standard way to encode function calls and/or results. WAVE is a human-oriented text encoding of Wasm Component Model values; designed to be consistent with the [WIT IDL format](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). 

Reviewer: Breaking this up might be useful -- it's quite dense. Some basic examples of WAVE might also help keep the user from clicking off and maybe going down a :rabbit2:  :hole:

#### Quotes

As shown above, the component's exported function name and mandatory parentheses are contained in one set of single quotes, i.e., `'get-answer()'`:

```console
$ wasmtime run --invoke 'get-answer()' target/wasm32-wasip1/debug/wasm_answer.wasm
```

The result from our correctly typed command above is as follows:

```console
42
```

#### Parentheses

Parentheses after the exported function's name are mandatory. The presence of the parenthesis `()` signifies function invocation, as opposed to the function name just being referenced. If your function takes a string argument, ensure that you contain your string in double quotes (inside the parentheses). For example:

```console
$ wasmtime run --invoke 'initialize("hello")' foo.wasm
```

If your exported function takes more than one argument, ensure that each argument is separated using a single comma `,` as shown below:

```console
$ wasmtime run --invoke 'initialize("Pi", 3.14)' foo.wasm
$ wasmtime run --invoke 'add(1, 2)' foo.wasm
```

## Recap: Wasm Modules versus Wasm Components

Let's wrap this article up with a recap to crystallize your knowledge.

### Wasm Modules

> 

Reviewer: This section speaks more to about how wamstime run *used* to work IMO, rather than WASM modules... 

Maybe the title should be something more like "Earlier wasmtime run support for modules"? Not sure

If we are not using the component model and just creating a binary executable, we use a simple command like `wasmtime run foo.wasm` (**without** WAVE syntax) to execute our the binary executable. This typically applies to modules, which export a `_start` function, or reactor modules, which can optionally export the `wasi:cli/run` interface—standardized to enable consistent execution semantics. 

Example of running a Wasm **module** that exports a raw function directly:

```console
$ wasmtime run --invoke initialize foo.wasm
```

### Wasm Components

> 

Reviewer: Since this is more about the addition of `wasmtime run --invoke`, maybe it should be titled that way?

Author: Updated this to align with your previous suggestion, updates in next push.

## Benefits and Usefulness

The addition of support for the run `--invoke` feature [for components](https://github.com/bytecodealliance/wasmtime/pull/10054) allows users to specify and execute exported functions from a Wasm component. This enables greater flexibility for testing, debugging, and integration. We now have the ability to perform the execution of arbitrary exported functions directly from the command line, this feature opens up a world of possibilities for integrating Wasm into modern development pipelines.

Here are some of the powerful workflows made possible by `--invoke`:

* Shell Scripting: Embed Wasm logic directly into Bash or Python scripts for seamless automation.
* CI/CD Pipelines: Validate components in GitHub Actions, GitLab CI, or other automation tools without embedding them in host applications.
* Cross-Language Testing: Quickly verify that interfaces match across implementations in Rust, JavaScript, and Python.
* Debugging: Inspect exported functions during development with ease.
* Microservices: Chain components in serverless workflows, such as compress → encrypt → upload, leveraging Wasm's modularity.

> **This evolution from monolithic Wasm modules to composable, CLI-friendly components exemplifies the versatility and power of WebAssembly in real-world scenarios.**

Reviewer: IMO (and forgive me if I made this point before and missed your follow up) -- I think this is really valuable as a "why you should keep reading" near the top of the article, below the description of the evolution

Author: Ah, I see. I will move this around. Please see changes in the next push.
